integtest/src/main/java/org/corfudb/universe/cluster/docker/DockerCluster.java

```diff
@@ -0,0 +1,204 @@
+import org.corfudb.universe.service.Service;
+import org.corfudb.universe.service.Service.ServiceParams;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.InetAddress;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.function.Function;
+
+import static org.corfudb.universe.node.CorfuServer.ServerParams;
+
+/**
+ * Represents Docker implementation of a {@link Cluster}.
+ */
+@Slf4j
+public class DockerCluster implements Cluster {
+    private static final FakeDns FAKE_DNS = FakeDns.getInstance().install();
+
+    @Getter
+    private final ClusterParams clusterParams;
+
+    private final DockerClient docker;
+    private final DockerNetwork network = new DockerNetwork();
+
+    private final ImmutableMap<String, Service> services;
+    private final Optional<String> clusterId;
+    private final Thread shutdownHook;
+
+    @Builder
+    public DockerCluster(ClusterParams clusterParams, DockerClient docker, ImmutableMap<String, Service> services,
+                         Optional<String> clusterId) {
+        this.clusterParams = clusterParams;
+        this.docker = docker;
+        this.services = services;
+        this.clusterId = clusterId;
+
+        shutdownHook = new Thread(this::shutdown);
+        Runtime.getRuntime().addShutdownHook(shutdownHook);
+    }
+
+
+    /**
+     * Deploy a cluster according to provided cluster parameter, docker client, docker network, and other cluster
+     * components.
+     * The instances of this class are immutable. In other word, when the state of an instance is changed a new
+     * immutable instance is provided.
+     *
+     * @return New immutable instance of docker cluster would be returned.
+     * @throws ClusterException this exception will be thrown if deploying a cluster is not successful
+     */
+    @Override
+    public DockerCluster deploy() throws ClusterException {
+        log.info("Deploying cluster");
+
+        if (!clusterId.isPresent()) {
+            network.setup();
+        }
+
+        Map<String, Service> servicesSnapshot = createAndDeployServices();
+
+        Optional<String> newClusterId = clusterId;
+        if (!newClusterId.isPresent()) {
+            newClusterId = Optional.of(UUID.randomUUID().toString());
+        }
+
+        DockerCluster cluster = DockerCluster.builder()
+                .clusterParams(clusterParams)
+                .docker(docker)
+                .services(ImmutableMap.copyOf(servicesSnapshot))
+                .clusterId(newClusterId)
+                .build();
+
+        Runtime.getRuntime().removeShutdownHook(shutdownHook);
+
+        return cluster;
+    }
+
+    public DockerCluster serviceManagmenet(String sericeNAme, Function<Service, Service> serviceFunction){
+        Service service = serviceFunction.apply(getService(sericeNAme));
+        services.put(service.getParams().getName(), service);
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Remove this use of "put"; it is deprecated. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)